src/utils/middleware.ts

```diff
@@ -1,7 +1,78 @@
-import { ResponseError } from '../pinecone-generated-ts-fetch';
+import { Middleware, ResponseError } from '../pinecone-generated-ts-fetch';
 import { handleApiError } from '../errors';
 
+const debugMiddleware: Middleware[] = [];
+
+const chalk = (str, color) => {
+  const colors = {
+    blue: '\x1b[34m',
+    red: '\x1b[31m',
+    green: '\x1b[32m',
+    yellow: '\x1b[33m',
+  };
+
+  return colors[color] + str + '\x1b[39m';
+};
+
+/**
+ * Enable the `PINECONE_DEBUG` environment variable to print the request and
+ * response bodies for each request.
+ *
+ * Api-Key headers will be redacted.
+ */
+if (process && process.env && process.env.PINECONE_DEBUG) {
+  const debugLogMiddleware = {
+    pre: async (context) => {
+      console.debug(
+        chalk(`>>> Request: ${context.init.method} ${context.url}`, 'blue')
+      );
+
+      const headers = JSON.parse(JSON.stringify(context.init.headers));
+      headers['Api-Key'] = '***REDACTED***';
+      console.debug(chalk(`>>> Headers: ${JSON.stringify(headers)}`, 'blue'));
+
+      if (context.init.body) {
+        console.debug(chalk(`>>> Body: ${context.init.body}`, 'blue'));
+      }
+      console.debug('');
+    },
+
+    post: async (context) => {
+      console.debug(chalk(`<<< Status: ${context.response.status}`, 'green'));
+      console.debug(
+        chalk(`<<< Body: ${await context.response.text()}`, 'green')
+      );
+      console.debug('');
+    },
+  };
+
+  debugMiddleware.push(debugLogMiddleware);
+}
+
+/**
+ * Enable the `PINECONE_DEBUG_CURL` environment variable to print the equivalent
+ * curl commands for each request. These commands will include the API key and
+ * other sensitive information, so be careful when using this option.
+ */
+if (process && process.env && process.env.PINECONE_DEBUG_CURL) {
+  const debugCurlMiddleware = {
+    post: async (context) => {
```

Comment:
**question:** I'm assuming the way it works on the OpenAPI side is if there are multiple `post` elements they're just executed sequentially on the specified event.

Was there documentation or anything you found for OpenAPI and how they support middleware, or was it all code spelunking? I'd be curious to take a look myself at some point.

Reply:
Just code spelunking. Relevant code is all in this [fetch method](https://github.com/pinecone-io/pinecone-ts-client/blob/main/src/pinecone-generated-ts-fetch/runtime.ts#L194). 

They are expecting `config.middleware` to be an array of middleware objects with optional `pre`, `onError`, and `post` callbacks. All are invoked, with the only caveat is if one of the middlewares throws an exception then nothing else gets called. This means you can only really have a single `onError` specified.